src/mechanics/elements/SpectralShapeFunctions.h

```diff
@@ -0,0 +1,248 @@
+    for (size_t j = 0; j < nodes.size(); j++)
+    {
+        for (size_t k = 0; k < nodes.size(); k++)
+        {
+            if (k != j)
+            {
+                double tmp = 1.;
+                for (size_t i = 0; i < nodes.size(); i++)
+                {
+                    if ((i != j) && (i != k))
+                    {
+                        tmp *= (x - nodes[i]);
+                    }
+                }
+                result[j] += tmp;
+            }
+        }
+        result[j] *= w[j];
+    }
+    return result;
+}
+}
+
+namespace ShapeFunctions2D
+{
+//! @brief Lobatto nodes for quad element (tensor product)
+//! @param order Polynomial order in 1D
+//! @return local node coordinates
+std::vector<Eigen::Vector2d> NodeCoordinatesQuadLobatto(int order)
+{
+    std::vector<double> nodes = ShapeFunctions1D::NodeCoordinatesTrussLobatto(order);
+    std::vector<Eigen::Vector2d> result;
+    for (size_t i=0; i<nodes.size(); i++)
+    {
+        for (size_t j=0; j<nodes.size(); j++)
+        {
+            result.push_back(Eigen::Vector2d({nodes[i],nodes[j]}));
+        }
+    }
+    return result;
+}
+
+//! @brief Value of Lagrange type shape functions with given 1D nodes (tensor product)
+//! @param x local coordinate where shapes are evaluated
+//! @param nodes local node coordinates
+//! @return shapes evaluated at x. Fulfill interpolation condition fi(xj) = delta_ij
+std::vector<double> ShapeFunctionsQuadLagrange(const Eigen::Vector2d x, const std::vector<double>& nodes)
+{
+    std::vector<double> Nx = ShapeFunctions1D::ShapeFunctionsTrussLagrange(x(0), nodes);
+    std::vector<double> Ny = ShapeFunctions1D::ShapeFunctionsTrussLagrange(x(1), nodes);
+
+    std::vector<double> result;
+    for (size_t j=0; j<nodes.size(); j++)
+    {
+        for (size_t i=0; i<nodes.size(); i++)
+        {
+            result.push_back(Nx[i] * Ny[j]);
+        }
+    }
+    return result;
+}
+
+std::vector<Eigen::Vector2d> DerivativeShapeFunctionsQuadLagrange(const Eigen::Vector2d x, const std::vector<double>& nodes)
+{
+    std::vector<double> DNx = ShapeFunctions1D::DerivativeShapeFunctionsTrussLagrange(x(0), nodes);
+    std::vector<double> DNy = ShapeFunctions1D::DerivativeShapeFunctionsTrussLagrange(x(1), nodes);
+
+    std::vector<double> Nx = ShapeFunctions1D::ShapeFunctionsTrussLagrange(x(0), nodes);
+    std::vector<double> Ny = ShapeFunctions1D::ShapeFunctionsTrussLagrange(x(1), nodes);
+
+    std::vector<Eigen::Vector2d> result;
+    for (size_t j=0; j<nodes.size(); j++)
+    {
+        for (size_t i=0; i<nodes.size(); i++)
+        {
+            result.push_back( Eigen::Vector2d({DNx[i]*Ny[j] , Nx[i]*DNy[j]})   );
+        }
+    }
+    return result;
+}
+
+}
+
+
+
+namespace ShapeFunctions3D
+{
+//! @brief Lobatto nodes for brick element (tensor product)
+//! @param order Polynomial order in 1D
+//! @return local node coordinates
+std::vector<Eigen::Vector3d> NodeCoordinatesBrickLobatto(int order)
```

Comment:
Is it possible to derive a `TensorProduct` abstraction? Ideally something like
~~~cpp
using NodeCoordinatesBrickLobatto = TensorProduct<3>(NodeCoordinatesTrussLobatto);
~~~
Similar for Quad and the other methods.

Reply:
This is probably a good idea. We talked about this when implementing the tensor product based integrationtypes and there such an abstraction would also be beneficial.